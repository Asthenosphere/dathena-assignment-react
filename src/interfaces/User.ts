export interface User {
  id: number | null;
  firstName: string | null;
  lastName: string | null;
  email: string | null;
  dob: Date | null;
}

export interface UserDetails {
  firstName: string;
  lastName: string;
  email: string;
  dob: Date;
}
