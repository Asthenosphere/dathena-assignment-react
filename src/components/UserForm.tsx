import React, { useState } from "react";
import { User } from "../interfaces/User";
import {
  IonButton,
  IonButtons,
  IonCol,
  IonContent,
  IonDatetime,
  IonGrid,
  IonHeader,
  IonInput,
  IonItem,
  IonLabel,
  IonList,
  IonModal,
  IonRow,
  IonTitle,
  IonToolbar,
} from "@ionic/react";

import { createUser } from "../services/userService";
import { isValidEmail } from "../utils/validationUtils";

interface UserFormProps {
  users: User[];
  isUserFormVisible: boolean;
  setIsUserFormVisible: (status: boolean) => void;
  userCallback: () => void;
  loadingCallback: (status: boolean) => void;
  alertCallback: (
    header: string,
    message: string,
    hasConfirm: boolean,
    confirmHandler: () => void
  ) => void;
}

const UserForm: React.FC<UserFormProps> = (props: UserFormProps) => {
  const {
    users,
    isUserFormVisible,
    setIsUserFormVisible,
    userCallback,
    loadingCallback,
    alertCallback,
  } = props;
  const [firstName, setFirstName] = useState<string>("");
  const [lastName, setLastName] = useState<string>("");
  const [email, setEmail] = useState<string>("");
  const [dob, setDob] = useState<string | undefined>(undefined);

  const handleSubmit = async () => {
    loadingCallback(true);
    const userObject = {
      firstName: firstName,
      lastName: lastName,
      email: email,
      dob: new Date(dob!).getTime(),
    };
    if (
      users.filter((x) => {
        return x.email === email;
      }).length > 0
    ) {
      loadingCallback(false);
      alertCallback(
        "Notice",
        "A user with the same email address has been found. Please try another email address.",
        false,
        () => {}
      );
      return;
    }
    try {
      await createUser(userObject);
      userCallback();
      setTimeout(() => {
        loadingCallback(false);
        alertCallback(
          "Success",
          `User ${firstName + " " + lastName} has been created successfully`,
          false,
          () => {}
        );
        setIsUserFormVisible(false);
        userCallback();
      }, 500);
    } catch (error) {
      loadingCallback(false);
      alertCallback("Error", error.message, false, () => {});
    }
  };

  const validateInputs = () => {
    return (
      firstName.length > 0 && lastName.length > 0 && isValidEmail(email) && dob
    );
  };

  return (
    <IonModal
      cssClass='modal'
      isOpen={isUserFormVisible}
      onDidDismiss={() => setIsUserFormVisible(false)}
    >
      <IonHeader>
        <IonToolbar>
          <IonTitle>New User</IonTitle>
          <IonButtons slot='end'>
            <IonButton onClick={() => setIsUserFormVisible(false)}>
              Close
            </IonButton>
          </IonButtons>
        </IonToolbar>
      </IonHeader>
      <IonContent>
        <IonGrid class='ion-padding-bottom ion-padding-horizontal'>
          <IonList>
            <IonItem>
              <IonLabel color='primary' position='floating'>
                First Name
              </IonLabel>
              <IonInput
                value={firstName}
                onIonChange={(event: CustomEvent) => {
                  setFirstName(event.detail.value);
                }}
              />
            </IonItem>
            <IonItem>
              <IonLabel color='primary' position='floating'>
                Last Name
              </IonLabel>
              <IonInput
                value={lastName}
                onIonChange={(event: CustomEvent) => {
                  setLastName(event.detail.value);
                }}
              />
            </IonItem>
            <IonItem>
              <IonLabel
                color={
                  email.length > 0 && !isValidEmail(email)
                    ? "danger"
                    : "primary"
                }
                position='floating'
              >
                Email
              </IonLabel>
              <IonInput
                value={email}
                onIonChange={(event: CustomEvent) => {
                  setEmail(event.detail.value);
                }}
              />
            </IonItem>
            <IonItem>
              <IonLabel color='primary' position='floating'>
                Date of Birth
              </IonLabel>
              <IonDatetime
                name='dob'
                displayFormat='DD MM YYYY'
                value={dob}
                onIonChange={(event: CustomEvent) => {
                  setDob(event.detail.value);
                }}
              />
            </IonItem>
          </IonList>
          <br />
          <IonRow class='ion-padding'>
            <IonCol>
              <IonButton
                expand='block'
                color='primary'
                shape='round'
                onClick={handleSubmit}
                disabled={!validateInputs()}
              >
                Create User
              </IonButton>
            </IonCol>
          </IonRow>
        </IonGrid>
      </IonContent>
    </IonModal>
  );
};

export default UserForm;
