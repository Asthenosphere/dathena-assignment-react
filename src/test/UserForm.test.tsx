import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";

import UserForm from "../components/UserForm";

import { User } from "../interfaces/User";

configure({ adapter: new Adapter() });

const users: User[] = [
  {
    id: 1,
    firstName: "John",
    lastName: "Doe",
    email: "john.doe@gmail.com",
    dob: new Date(1595493789),
  },
];

it("renders user form correctly", () => {
  const form = shallow(
    <UserForm
      users={users}
      isUserFormVisible={true}
      setIsUserFormVisible={() => {}}
      userCallback={() => {}}
      loadingCallback={() => {}}
      alertCallback={(
        header: string,
        message: string,
        hasConfirm: boolean,
        confirmHandler: () => void
      ) => {}}
    />
  );
  expect(form.getElements()).toMatchSnapshot();
});
