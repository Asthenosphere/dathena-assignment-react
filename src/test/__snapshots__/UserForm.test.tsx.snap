// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders user form correctly 1`] = `
Array [
  <ForwardRef
    cssClass="modal"
    isOpen={true}
    onDidDismiss={[Function]}
  >
    <ForwardRef(IonHeader)>
      <ForwardRef(IonToolbar)>
        <ForwardRef(IonTitle)>
          New User
        </ForwardRef(IonTitle)>
        <ForwardRef(IonButtons)
          slot="end"
        >
          <ForwardRef(IonButton)
            onClick={[Function]}
          >
            Close
          </ForwardRef(IonButton)>
        </ForwardRef(IonButtons)>
      </ForwardRef(IonToolbar)>
    </ForwardRef(IonHeader)>
    <ForwardRef(IonContent)>
      <ForwardRef(IonGrid)
        class="ion-padding-bottom ion-padding-horizontal"
      >
        <ForwardRef(IonList)>
          <ForwardRef(IonItem)>
            <ForwardRef(IonLabel)
              color="primary"
              position="floating"
            >
              First Name
            </ForwardRef(IonLabel)>
            <ForwardRef(IonInput)
              onIonChange={[Function]}
              value=""
            />
          </ForwardRef(IonItem)>
          <ForwardRef(IonItem)>
            <ForwardRef(IonLabel)
              color="primary"
              position="floating"
            >
              Last Name
            </ForwardRef(IonLabel)>
            <ForwardRef(IonInput)
              onIonChange={[Function]}
              value=""
            />
          </ForwardRef(IonItem)>
          <ForwardRef(IonItem)>
            <ForwardRef(IonLabel)
              color="primary"
              position="floating"
            >
              Email
            </ForwardRef(IonLabel)>
            <ForwardRef(IonInput)
              onIonChange={[Function]}
              value=""
            />
          </ForwardRef(IonItem)>
          <ForwardRef(IonItem)>
            <ForwardRef(IonLabel)
              color="primary"
              position="floating"
            >
              Date of Birth
            </ForwardRef(IonLabel)>
            <ForwardRef(IonDatetime)
              displayFormat="DD MM YYYY"
              name="dob"
              onIonChange={[Function]}
            />
          </ForwardRef(IonItem)>
        </ForwardRef(IonList)>
        <br />
        <ForwardRef(IonRow)
          class="ion-padding"
        >
          <ForwardRef(IonCol)>
            <ForwardRef(IonButton)
              color="primary"
              disabled={true}
              expand="block"
              onClick={[Function]}
              shape="round"
            >
              Create User
            </ForwardRef(IonButton)>
          </ForwardRef(IonCol)>
        </ForwardRef(IonRow)>
      </ForwardRef(IonGrid)>
    </ForwardRef(IonContent)>
  </ForwardRef>,
]
`;
